### ISSUE TITLE
Proposal: disallow regular expressions with the global match flag `g` at the module level

### ISSUE THREAD

**Issue by lencioni** [eslint/eslint#8088]
**Please describe what the rule should do:**

Regular expressions with the `g` flag maintain state, which can cause confusion and subtle bugs if the regex is defined at the module level. It would be nice if there was an eslint rule to prevent this.

It might also be worth flagging regexes with the `g` flag in the global scope.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Bad:

```js
const foo = /./g;
```
```js
const foo = /./gi;
```
```js
const bar = new RegExp('.', 'g');
```
```js
const baz = new RegExp(/./, 'g');
```

Okay:

```js
function() {
  const foo = /./g;
}
```
```js
function() {
  const bar = new RegExp('.', 'g');
}
```
```js
const foo = /./;
```
```js
const bar = new RegExp('.');
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a source of bugs in a core part of the language.

cc @ljharb

**Comment by ljharb** [comment]
This bit me specifically because I *thought* about `lastIndex`, and intentionally used the `g` flag, thinking it would protect me from the mutation - when in fact, it was the reverse.

Essentially this rule should warn whenever a regex literal (or `new RegExp`) with the global flag is re-used across scopes.

**Comment by not-an-aardvark** [comment]
Thanks for the rule request.

My initial impression is that this rule is too broad to be a good fit for core. In other words, it would catch some bugs, but it would also report a lot of cases which are not bugs. A given regex could be intentionally mutable (e.g. if it's used in a `while (regex.exec())` loop), or it might not ever be observably mutated (e.g. if it's only used in `String#replace`). I agree that immutability is a good thing to have in general, but given that mutation is sometimes necessary, I'm not sure core rules should be enforcing immutability. This might be a better fit for something like [eslint-plugin-fp](https://github.com/jfmengels/eslint-plugin-fp) instead.

**Comment by lencioni** [comment]
@jfmengels what do you think?

**Comment by ljharb** [comment]
@not-an-aardvark the OP asked for "module level" - do you expect intentional mutable regexes to occur outside a function?

**Comment by not-an-aardvark** [comment]
> the OP asked for "module level" - do you expect intentional mutable regexes to occur outside a function?

Yes, I suspect this is relatively common. For example, one could use a pattern like this:

```js
const WHITESPACE = /\s/g;

function replaceWithSpaces(text) {
  return text.replace(WHITESPACE, ' ');
}

function replaceWithTabs(text) {
  return text.replace(WHITESPACE, '\t');
}
```

Or (I suspect this is less common) someone could use an `exec` loop at the top level:

```js
const file = fs.readFileSync('input-data.txt');

const WORD_MATCHER = /[^s]+/g;
let match;

while ((match = WORD_MATCHER.exec(file))) {
  // ...
}
```

**Comment by ljharb** [comment]
Certainly in those examples (assuming the while loop always completes normally) there's no hazard, but that being true seems like an advanced enough case that it'd warrant an eslint override comment; since the potential there for error is very high.

The point of a rule isn't necessarily to cover all cases, it's often to do the safe/best thing in the common case.

**Comment by not-an-aardvark** [comment]
> The point of a rule isn't necessarily to cover all cases, it's often to do the safe/best thing in the common case.

Definitely agree. I'm just concerned about whether the "common case" for this rule (accidental mutation) for this rule is actually much more common than the safe case. Sure, people can use override comments for exceptional cases, but if they need to be used for a very significant percentage of errors reported by a rule, then the rule can be sort of irritating to use.

**Comment by jfmengels** [comment]
I didn't know about the internal mutation. Can someone give an example of where this would bite you? 

Is the problem only when having a regex with the `g` flag defined at module level, or is it simply at any level but reusing it multiple times? If the number of false positives are pretty low, I think this should go into core, otherwise I agree that this should go in a plugin (like @not-an-aardvark said). Also, depending on this, I wouldn't mind having it in eslint-plugin-fp, but I'm not sure it's the best fit for it. IMO, this should rather be in a plugin dedicated to regexes, but there are no popular ones unfortunately.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
